### PR TITLE
Implement configurable acquisition time for ADS114s0x

### PR DESCRIPTION
**Pull request by benediktibk**
Implement a configurable acquisition time for the ADS114s0x.

### CODE REVIEW

File: drivers/adc/adc_ads114s0x.c [hunk 1]
```diff
@@ -512,6 +534,69 @@ static int ads114s0x_channel_setup(const struct device *dev,
...
+	 * within here for microsecond units. Use Tick units and allow the user to
+	 * specify the ODR directly.
+	 */
+	if (acquisition_time_value != ADC_ACQ_TIME_DEFAULT &&
+	    acquisition_time_unit != ADC_ACQ_TIME_TICKS) {
+		LOG_ERR("invalid acquisition time %i", channel_cfg->acquisition_time);
+		return -EINVAL;
+	}
+
+	if (acquisition_time_value == ADC_ACQ_TIME_DEFAULT) {
```
Comment (anangl) [inline drivers/adc/adc_ads114s0x.c:548]:
```suggestion
	if (channel_cfg->acquisition_time == ADC_ACQ_TIME_DEFAULT) {
```

Comment (benediktibk) [reply drivers/adc/adc_ads114s0x.c]: Thanks, done.

File: drivers/adc/adc_ads114s0x.c [hunk 2]
```diff
@@ -76,6 +76,26 @@ enum ads114s0x_register {
 	ADS114S0X_REGISTER_GPIOCON = 0x11,
 };
 
+/* available data rates described as samples per second, should be used with the time unit ticks */
+enum ads114s0x_data_rate {
+	ADS114S0X_CONFIG_DR_2_5 = 1, /* must not be equal to ADC_ACQ_TIME_DEFAULT */
```
Comment (anangl) [inline drivers/adc/adc_ads114s0x.c:81]: This definition is internal to the driver. How will the user know what value to use in the call to `ADC_ACQ_TIME()`? Wouldn't it be better to define such symbols in `dt-bindings/adc`?

File: drivers/adc/adc_ads114s0x.c [hunk 3]
```diff
@@ -76,6 +76,26 @@ enum ads114s0x_register {
...
+	ADS114S0X_CONFIG_DR_100,
+	ADS114S0X_CONFIG_DR_200,
+	ADS114S0X_CONFIG_DR_400,
+	ADS114S0X_CONFIG_DR_800,
+	ADS114S0X_CONFIG_DR_1000,
+	ADS114S0X_CONFIG_DR_2000,
+	ADS114S0X_CONFIG_DR_4000,
+};
+
+BUILD_ASSERT(ADC_ACQ_TIME_DEFAULT != ADS114S0X_CONFIG_DR_2_5);
```
[the remark this answers is not in the document]
Comment (benediktibk) [reply drivers/adc/adc_ads114s0x.c:97]: done

File: drivers/adc/adc_ads114s0x.c (hunk 2 above)
Comment (benediktibk) [reply drivers/adc/adc_ads114s0x.c]: I absolutely agree with you, but I basically just stole the way to implement this from here:
https://github.com/zephyrproject-rtos/zephyr/blob/ea2aac6e5131bf916abd1e37334e94cc3481e1a8/drivers/adc/adc_ads1x1x.c#L54

Comment (benediktibk) [reply drivers/adc/adc_ads114s0x.c]: I just had a rather similar discussion in https://github.com/zephyrproject-rtos/zephyr/pull/56234#discussion_r1266990855 and my conclusion is that most drivers seem to expose this kind of information via a header. Therefore, I will go ahead and move this into a driver specific header.

Comment (benediktibk) [reply drivers/adc/adc_ads114s0x.c]: done

Comment (benediktibk) [reply drivers/adc/adc_ads114s0x.c]: I just realized you mentioned dt-bindings/adc, moved the enum definition there.

**Review by anangl — DISMISSED** [review]

File: include/zephyr/dt-bindings/adc/ads114s0x_adc.h [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+ * Copyright (c) 2023 SILA Embedded Solutions GmbH
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DT_BINDINGS_ADC_ADS114S0X_ADC_H_
+#define ZEPHYR_INCLUDE_DT_BINDINGS_ADC_ADS114S0X_ADC_H_
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/gpio.h>
```
[the remark this answers is not in the document]
Comment (benediktibk) [reply include/zephyr/dt-bindings/adc/ads114s0x_adc.h]: done

File: include/zephyr/dt-bindings/adc/ads114s0x_adc.h [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+/*
+ * These are the available data rates described as samples per second. They
+ * can be used with the time unit ticks for the acquisition time.
+ */
+enum ads114s0x_data_rate {
```
[the remark this answers is not in the document]
Comment (benediktibk) [reply include/zephyr/dt-bindings/adc/ads114s0x_adc.h]: done

File: drivers/adc/adc_ads114s0x.c [hunk 6]
```diff
@@ -607,6 +610,69 @@ static int ads114s0x_channel_setup(const struct device *dev,
...
+		LOG_ERR("invalid acquisition time %i", channel_cfg->acquisition_time);
+		return -EINVAL;
+	}
+
+	if (channel_cfg->acquisition_time == ADC_ACQ_TIME_DEFAULT) {
+		ADS114S0X_REGISTER_DATARATE_DR_SET(data_rate, 0b0100);
+	} else {
+		switch (acquisition_time_value) {
+		case ADS114S0X_CONFIG_DR_2_5:
+			ADS114S0X_REGISTER_DATARATE_DR_SET(data_rate, 0b0000);
```
Comment (aasinclair) [inline drivers/adc/adc_ads114s0x.c:629]: As you have defined the available data rates to align with the register settings, couldn't all these case entries be replaced with:
`ADS114S0X_REGISTER_DATARATE_DR_SET(data_rate, acquisition_time_value);`

Comment (benediktibk) [reply drivers/adc/adc_ads114s0x.c:629]: I already thought about the exact same thing, but was kind of uncertain if I should keep it in such an explicit form or not. Technically there are two possible values for 4000 SPS (0b1101 and 0b1110), for instance. But I get the point of reducing the amount of code, I will change it.

Comment (benediktibk) [reply drivers/adc/adc_ads114s0x.c]: done

**Comment by benediktibk** [comment]
Looks like the build error is caused by something else:

    FAILED: zephyr/zephyr.ri /__w/zephyr/zephyr/twister-out/intel_adsp_cavs25/tests/boards/intel_adsp/ssp/boards.intel_adsp.ssp/zephyr/zephyr.ri 
    cd /__w/zephyr/zephyr/twister-out/intel_adsp_cavs25/tests/boards/intel_adsp/ssp/boards.intel_adsp.ssp/zephyr/soc/xtensa/intel_adsp/common && west sign --if-tool-available --tool rimage --build-dir /__w/zephyr/zephyr/twister-out/intel_adsp_cavs25/tests/boards/intel_adsp/ssp/boards.intel_adsp.ssp
    usage: west [-h] [-z ZEPHYR_BASE] [-v] [-V] <command> ...
    west: error: argument <command>: invalid choice: 'sign' (choose from 'init', 'update', 'list', 'manifest', 'diff', 'status', 'forall', 'help', 'config', 'topdir', 'selfupdate')
    ninja: build stopped: subcommand failed.

**Review by aasinclair — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]

**Review by anangl — APPROVED** [review]

**Pull request merged**